File: golang/143_reorder_list.go
package main

// ListNode Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func reorderList(head *ListNode) {
	if head == nil {
		return
	}
	list := []*ListNode{}
	tmp := head

	for tmp != nil {
		list = append(list, tmp)
		tmp = tmp.Next
	}

	h := head
	for i := len(list) - 1; i > (len(list)-1)/2; i-- {
		next := h.Next
		h.Next = list[i]
		h.Next.Next = next
		h = next
	}
	h.Next = nil
}

File: python/143_reorder_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if head == None:
            return head
        temp = head.next
        doubleEnded = collections.deque([])
        while temp:
            doubleEnded.append(temp)
            temp = temp.next
        popRight = True
        temp = head
        while len(doubleEnded) > 0:
            if popRight:
                temp2 = doubleEnded.pop()
                temp.next = temp2
                temp = temp2
                popRight = False
            else:
                temp2 = doubleEnded.popleft()
                temp.next = temp2
                temp = temp2
                popRight = True
        temp.next = None